keps/sig-node/4781-kublet-restart-pod-status/README.md

```diff
@@ -0,0 +1,581 @@
+It's essential to implement strategies to ensure that the service states accurately reflect their operational status, even during kubelet interruptions.
+
+### Goals
+
+- Ensure consistency in container start and ready states across kubelet restarts.
+- Minimize unnecessary service disruptions caused by temporary ready state changes.
+
+### Non-Goals
+
+- If the kubelet fails to renew its lease beyond the nodeMonitorGracePeriod due to an excessively long restart interval, the Ready status of the containers in the pods on the node will be set to false. In this situation, we should not manually set the Ready status back to true. Instead, it should remain false, waiting for the probe to execute again and restore it.
+- Modify the fundamental logic of how readiness probes work.
+
+## Proposal
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+As a user of Kubernetes, I want the container's Ready state to remain consistent across kubelet restarts so that my services do not experience unnecessary downtime.
+However, currently, a kubelet restart causes a brief "Not Ready" storm, where the state of all Pods is set to Not Ready, impacting the availability of my services.
+
+### Risks and Mitigations
+
+##### Inconsistency with other Kubernetes components
+If other parts of Kubernetes (e.g., the API server, controllers) expect certain behavior regarding container readiness states, these changes might cause inconsistencies.
+
+##### Delayed Health Check Updates
+By preserving the old state without immediate health checks, there is a delay in recognizing containers that have become unhealthy during or after kubelet's downtime. Services relying on Pod readiness for service discovery might continue directing traffic to Pods with containers that are no longer healthy but are still reported as Ready.
+We plan to immediately trigger a probe after that to reduce the risk caused by such delays.
+
+## Design Details
+
+1. We will retrieve the `Started` field from the container status in the Pod via the API server. After the Kubelet restarts, during the first entry into `SyncPod`, we will propagate this value to the newly generated container status.
+
+2. We ensure that if the `Started` field in the container status is true, the container is considered started (since the startupProbe only runs during container startup and will not execute again once completed).
+
+3. If the Kubelet restart occurs within the `nodeMonitorGracePeriod` and the Pod’s Ready condition is set to false, we will set the container’s ready status to false. It will remain in this state until subsequent probes reset it to true.
```

Comment:
> If the Kubelet restart occurs within the `nodeMonitorGracePeriod`

Does this mean the case where the kubelet has been down for longer than `nodeMonitorGracePeriod` and then restarts afterward?

So basically, the Node Lifecycle Controller notices that the Lease hasn’t been updated past `nodeMonitorGracePeriod`, marks the Node as NotReady, and flips the Pods’ Ready condition to False. After the kubelet restarts, it fetches the Pod info for its own Node from the API server, and the prober manager simply carries over that Ready condition value, right?